scheds/rust/scx_layered/src/bpf/main.bpf.c

```diff
@@ -848,13 +905,28 @@ s32 pick_idle_cpu(struct task_struct *p, s32 prev_cpu,
 	 */
 	if (READ_ONCE(layer->check_no_idle)) {
 		bool has_idle;
-
 		cpumask = scx_bpf_get_idle_cpumask();
 
-		if (layer->kind == LAYER_KIND_CONFINED)
+		if (layer->kind == LAYER_KIND_CONFINED) {
 			has_idle = bpf_cpumask_intersects(layered_cpumask, cpumask);
-		else
-			has_idle = bpf_cpumask_intersects(p->cpus_ptr, cpumask);
+		} else {
+			maybe_refresh_layered_cpus_unprotected(p, taskc);
+			/*
+			 * Use the task's idle unprotected mask if available, otherwise
+			 * use the global one.
+			 */
+			unprot_mask = taskc->layered_unprotected_mask;
+			if (!unprot_mask)
+				unprot_mask = unprotected_cpumask;
+
+			if (unlikely(!unprot_mask)) {
+				scx_bpf_error("unprotected_cpumask not initialized");
+				scx_bpf_put_idle_cpumask(cpumask);
+				return -1;
+			}
+
+			has_idle = bpf_cpumask_intersects(cast_mask(unprot_mask), cpumask);
```

Comment:
and then we're saying "if `(p->cpus_allowed & unprotected_mask) & idle_cpumask` is empty, we have no idle CPUs". However, that's not quite the right condition. The right condition would be `(p->cpus_allowed & (layered->cpus | unprotected_cpumask)` because the task has access to both the layer's owned CPUs and unprotected CPUs.